### PR TITLE
Refactor interactions render edit page

### PR DESCRIPTION
**Pull request by yeahfro**
This changes focuses around refactor `renderEditPage` for interactions. We ultimately want to mount interactions as a sub app within other apps.

Notes:
- a temporary route has been created within `router.js` of interactions. This demonstrates the setting of `res.locals.companies` `res.locals.returnLink` depending on the app
- I feel uncomfortable about `getInteractionTypeAndService` - any thoughts?

screen shot 2017-10-06 at 09 18 41

### CODE REVIEW

File: src/apps/interactions/middleware/details.js [hunk 1]
```diff
@@ -29,6 +35,48 @@ async function postDetails (req, res, next) {
...
+  try {
+    res.locals.interaction = await fetchInteraction(req.session.token, interactionId)
+    res.locals.interactionViewRecord = transformInteractionResponseToViewRecord(res.locals.interaction)
+    next()
+  } catch (err) {
+    next(err)
+  }
+}
+
+async function getContactsForCompanyDetails (req, res, next) {
```
Comment (ztolley) [inline src/apps/interactions/middleware/details.js]: Feels like a weird name, Is there a middleware that could be added to companies to get contacts and then call it here?

File: src/apps/interactions/middleware/details.js [hunk 2]
```diff
@@ -29,6 +35,48 @@ async function postDetails (req, res, next) {
...
+  try {
+    res.locals.contacts = await getContactsForCompany(req.session.token, res.locals.company.id)
+    next()
+  } catch (err) {
+    next(err)
+  }
+}
+
+async function getAdviserDetails (req, res, next) {
```
Comment (ztolley) [inline src/apps/interactions/middleware/details.js]: maybe call it just getAdvisers and also maybe as a middleware under advisers app?

File: src/apps/interactions/middleware/details.js [hunk 3]
```diff
@@ -29,6 +35,48 @@ async function postDetails (req, res, next) {
...
+  try {
+    res.locals.advisers = await getAdvisers(req.session.token)
+    next()
+  } catch (err) {
+    next(err)
+  }
+}
+
+async function getInteractionTypeAndService (req, res, next) {
```
Comment (ztolley) [inline src/apps/interactions/middleware/details.js]: getServicesForInteractionType? dunno

File: src/apps/interactions/router.js [hunk 4]
```diff
@@ -20,9 +26,26 @@ router
...
-router.route(['/create/2', '/:id/edit'])
-  .post(postDetails, renderEditPage)
-  .get(renderEditPage)
+router.route(['/create/2', '/:interactionId/edit'])
+  .post(
+    postDetails,
+    renderEditPage,
+  )
+  .get(
+    async function (req, res, next) {
```
Comment (ztolley) [inline src/apps/interactions/router.js]: You're moving this into a separate file right?

File: src/apps/interactions/transformers.js [hunk 5]
```diff
@@ -72,6 +66,29 @@ function transformInteractionToListItem ({
   }
 }
 
+function transformInteractionResponseToViewRecord ({
```
Comment (ztolley) [inline src/apps/interactions/transformers.js]: I take this will be replaced by the view transformer I'm about to merge?

File: src/apps/interactions/middleware/details.js (hunk 2 above)
Comment (yeahfro) [reply src/apps/interactions/middleware/details.js]: @ztolley that's what I did yesterday and we talked about premature optimisation!

File: src/apps/interactions/middleware/details.js (hunk 3 above)
Comment (yeahfro) [reply src/apps/interactions/middleware/details.js]: @ztolley that would imply that the `services` are retrieved based upon the `interactionType`

File: src/apps/interactions/router.js (hunk 4 above)
Comment (yeahfro) [reply src/apps/interactions/router.js]: @ztolley this implementation will be deleted as interactions will no longer be mounted as a root app. This was just an example.

File: src/apps/interactions/transformers.js (hunk 5 above)
Comment (yeahfro) [reply src/apps/interactions/transformers.js]: @ztolley yeah I will sort out the merge conflicts shortly

File: src/apps/interactions/middleware/details.js (hunk 1 above)
Comment (yeahfro) [reply src/apps/interactions/middleware/details.js]: @ztolley yes I feel uncomfortable with this and `getAdviserDetails` but I would anticipate we can spot patterns and refactor over time.

File: src/apps/interactions/router.js (hunk 4 above)
Comment (ztolley) [reply src/apps/interactions/router.js]: Now it's going back into the top level nav then we need to support it both as a sub and root app (though wont have add in root list)

File: src/apps/interactions/router.js [hunk 6]
```diff
@@ -20,9 +26,26 @@ router
...
+      // set company
+      const companyRepo = require('../companies/repos')
+      res.locals.company = await companyRepo.getDitCompany(req.session.token, req.query.company)
+
+      // set return link
+      res.locals.returnLink = `/companies/${req.query.company}/interactions`
```
Comment (teneightfive) [inline src/apps/interactions/router.js]: `returnUrl`? I've used that variable name within OMIS and seems clearer to me.

File: src/apps/interactions/router.js (hunk 4 above)
Comment (teneightfive) [reply src/apps/interactions/router.js]: I've been thinking about how we handle returning to a particular spot and wonder if we could have a higher level middleware across all entities that accepts a `returnUrl` query string param and returns a successful submit back there and sets the cancel/back url.

Would solve problems like adding a contact in the middle of creating an OMIS order for example.

File: src/apps/interactions/transformers.js [hunk 7]
```diff
@@ -4,32 +4,26 @@ const dateFns = require('date-fns')
...
-    interaction_type: get(interaction_type, 'id'),
-    company: get(company, 'id'),
-    subject,
-    notes,
-    contact: get(contact, 'id'),
-    date: {
-      day: dateFns.format(date, 'DD'),
-      month: dateFns.format(date, 'MM'),
-      year: dateFns.format(date, 'YYYY'),
+function transformInteractionResponseToForm (props = {}) {
```
Comment (teneightfive) [inline src/apps/interactions/transformers.js]: What's the reason for changing this format? Are we only overriding certain properties with this transformer?

File: src/apps/interactions/middleware/details.js (hunk 2 above)
Comment (jim68000) [reply src/apps/interactions/middleware/details.js]: YAGNI,Y

File: src/apps/interactions/router.js (hunk 6 above)
Comment (jim68000) [reply src/apps/interactions/router.js]: 👍

Comment (yeahfro) [reply src/apps/interactions/router.js]: @teneightfive not sure, the form builder expects a `returnLink`. Shall we change it everywhere?

File: src/apps/interactions/controllers/edit.js [hunk 8]
```diff
@@ -1,57 +1,43 @@
...
-            interaction_type: formData.interaction_type,
-            company: formData.company,
-          },
-        }
-      ),
-      formData,
-      get(res.locals, 'form.errors.messages'),
-    )
+    const interactionData = transformInteractionResponseToForm(res.locals.interaction)
+    const dit_adviser = interactionData.dit_adviser || get(req, 'session.user')
```
Comment (feedmypixel) [inline src/apps/interactions/controllers/edit.js]: can this const be camelCase like the otehrs?

File: src/apps/interactions/controllers/edit.js [hunk 9]
```diff
@@ -1,57 +1,43 @@
...
+    const date = interactionData.date || transformDateStringToDateObject(new Date())
```
Comment (feedmypixel) [inline src/apps/interactions/controllers/edit.js]: I'd give this a better name `interactionDate` maybe?

File: src/apps/interactions/controllers/edit.js [hunk 10]
```diff
@@ -1,57 +1,43 @@
...
+    const mergedData = pickBy(merge({}, interactionData, { dit_adviser, date }, res.locals.requestBody))
+    const advisers = get(res.locals, 'advisers.results')
```
Comment (feedmypixel) [inline src/apps/interactions/controllers/edit.js]: do these need to be `const`? they are only used in one place Lines `26`, `27`, `28`

File: src/apps/interactions/controllers/edit.js [hunk 11]
```diff
@@ -1,57 +1,43 @@
...
-          },
-        }
-      ),
-      formData,
-      get(res.locals, 'form.errors.messages'),
-    )
+    const interactionData = transformInteractionResponseToForm(res.locals.interaction)
+    const dit_adviser = interactionData.dit_adviser || get(req, 'session.user')
+    const date = interactionData.date || transformDateStringToDateObject(new Date())
+    const mergedData = pickBy(merge({}, interactionData, { dit_adviser, date }, res.locals.requestBody))
```
Comment (feedmypixel) [inline src/apps/interactions/controllers/edit.js]: rather than naming this by the action (`mergedData`) can it be named by what it actually is?

File: src/apps/interactions/controllers/edit.js [hunk 12]
```diff
@@ -1,57 +1,43 @@
...
+        get(res.locals, 'form.errors.messages'),
+      )
 
     res
-      .breadcrumb(pageTitle)
+      .breadcrumb(`${interactionId ? 'Edit' : 'Add'} interaction for ${res.locals.company.name}`)
       .render('interactions/views/edit', {
-        editInteractionForm,
-        interactionTypeLabel: interaction.interaction_type.name,
+        interactionForm: assign(interactionForm, {
```
Comment (feedmypixel) [inline src/apps/interactions/controllers/edit.js]: it ok that `interactionForm` is being muted here? Feels a little more functional to add `{}` as first argument. More preference though, so your call.

File: src/apps/interactions/middleware/details.js [hunk 13]
```diff
@@ -1,7 +1,13 @@
...
-
-const { transformInteractionFormBodyToApiRequest } = require('../transformers')
-const { saveInteraction } = require('../repos')
+const {
+  transformInteractionResponseToViewRecord,
+  transformInteractionFormBodyToApiRequest,
+} = require('../transformers')
+const { fetchInteraction, saveInteraction } = require('../repos')
+const metaDataRepository = require('../../../lib/metadata')
+const { find } = require('lodash')
```
Comment (feedmypixel) [inline src/apps/interactions/middleware/details.js]: can be combined with the above `require('lodash')`

File: src/apps/interactions/middleware/details.js (hunk 2 above)
Comment (feedmypixel) [reply src/apps/interactions/middleware/details.js]: @ztolley `getAdvisers` is already a method (used within this method) hence it was originally called `getAdviserDetails`

File: src/apps/interactions/middleware/details.js (hunk 1 above)
Comment (feedmypixel) [reply src/apps/interactions/middleware/details.js]: surely `getCompanyDetails`?

File: src/apps/interactions/middleware/details.js (hunk 3 above)
Comment (feedmypixel) [reply src/apps/interactions/middleware/details.js]: `getInteractionDetails`?

File: src/apps/interactions/repos.js [hunk 14]
```diff
@@ -20,6 +20,10 @@ function saveInteraction (token, interaction) {
   return authorisedRequest(token, options)
 }
 
+function fetchInteraction (token, id) {
+  return authorisedRequest(token, `${config.apiRoot}/v3/interaction/${id}`)
```
Comment (feedmypixel) [inline src/apps/interactions/repos.js]: the use of `fetch` feels wrong here. We use `get` all over the application. `fetch` to me implies a http call

File: src/apps/interactions/transformers.js (hunk 7 above)
Comment (feedmypixel) [reply src/apps/interactions/transformers.js]: I'd also stick with the format we have elsewhere for transformers

File: src/apps/interactions/transformers.js [hunk 15]
```diff
@@ -84,4 +101,5 @@ module.exports = {
   transformInteractionResponseToForm,
   transformInteractionToListItem,
   transformInteractionFormBodyToApiRequest,
+  transformInteractionResponseToViewRecord,
```
Comment (feedmypixel) [inline src/apps/interactions/transformers.js]: is `Record` needed at the end here?

File: test/unit/apps/interactions/controllers/edit.test.js [hunk 16]
```diff
@@ -1,166 +1,127 @@
...
 
-    context('when success', () => {
-      it('should call createBlankInteractionForCompany for company', async () => {
-        this.req.query.company = 'Hooli Inc'
-        const expectedArgs = ['abcd', { name: 'Spencer' }, 'tel', 'Hooli Inc']
+    it('should render the interaction page with an interaction type label', async () => {
+      await this.controller.renderEditPage(this.req, this.res, this.next)
 
-        await this.edit.renderEditPage(this.req, this.res, this.nextSpy)
+      const actual = this.res.render.getCall(0).args[1].interactionForm.interactionTypeLabel
```
Comment (feedmypixel) [inline test/unit/apps/interactions/controllers/edit.test.js]: I don't feel testing the labels here is the right thing to be testing in a unit test. 

IMO you should unit test the transformer to make sure it gives you back what is expected. You should unit test the controller to make sure it goes in the right direction. I don't feel unit testing the controllers and the values that the controllers passes on gives us much value here. It does give us a lot of maintenance.

- Controllers should just be "this goes in the direction I expect it to" by testing `next()`/`render()`
- Transformers can be tested by saying "I give the transformer this and I expect this"

This means we don't then test controllers for all the different values being sent through the one transformer.

As always open to discussion on this

**Comment by feedmypixel** [comment]
@yeahfro There have a been a lot of tests removed in this PR. Can we wait to remove tests until we have had the chat on Thursday. I want to make sure we are all on the same page before we start removing swathes of the tests.

File: test/unit/apps/interactions/controllers/edit.test.js (hunk 16 above)
Comment (ztolley) [reply test/unit/apps/interactions/controllers/edit.test.js]: But I've always thought we should test that the controller returns the result of a transformation, to make sure it calls it and that it sends it back.

I think if we move to a integration approach then this would an end to this discussion :)

File: src/apps/interactions/transformers.js (hunk 15 above)
Comment (ztolley) [reply src/apps/interactions/transformers.js]: I think that's the naming convention we've followed recently after being suggested by Tyom

File: src/apps/interactions/middleware/details.js (hunk 1 above)
Comment (yeahfro) [reply src/apps/interactions/middleware/details.js]: @feedmypixel this is only getting the contacts for the company

File: src/apps/interactions/router.js (hunk 6 above)
Comment (ztolley) [reply src/apps/interactions/router.js]: Consistency is good

File: src/apps/interactions/middleware/details.js (hunk 3 above)
Comment (yeahfro) [reply src/apps/interactions/middleware/details.js]: @feedmypixel there is already a `getInteractionDetails` which is used by the router for param `interactionId`. This function needs to be added as middleware even when there is not an `interactionId`

**Comment by feedmypixel** [comment]
> @yeahfro There have a been a lot of tests removed in this PR. Can we wait to remove tests until we have had the chat on Thursday. I want to make sure we are all on the same page before we start removing swathes of the tests.

sorry to clarify, if they need removing as part of this work go for it. If they are being removed re: integration/ testing chat lets hold off till then

File: test/unit/apps/interactions/controllers/edit.test.js (hunk 16 above)
Comment (feedmypixel) [reply test/unit/apps/interactions/controllers/edit.test.js]: > But I've always thought we should test that the controller returns the result of a transformation, to make sure it calls it and that it sends it back.

And that sir is an integration test. Which yes we should be doing but in integration tests not in unit tests. TBH (more) 😉

Comment (ztolley) [reply test/unit/apps/interactions/controllers/edit.test.js]: I test I call something and that I used the correct thing. I know we disagree on this but it's fine, we'll move to integration tests and agree on the pattern there.

File: src/apps/interactions/router.js (hunk 4 above)
Comment (ztolley) [reply src/apps/interactions/router.js]: I'm still torn between going to 
`/interactions/:interactionId` and `/contact/:contactId/interaction/:interactionId`

Are we trying to make it harder than it needs to be?

Should we not just go with the former, and use something like return URL for a back button? Even add a returnTitle so you would have a link 'Return to contact' ?

Comment (feedmypixel) [reply src/apps/interactions/router.js]: @ztolley think of it as a signpost:
`/interactions/:interactionId`
`/contact/:contactId/interaction/:interactionId` (edited)
which 1 out of the above 2 tells you what it is and where it is? which 1 just tells you what it is?

Comment (ztolley) [reply src/apps/interactions/router.js]: Users don't care about url's though

Comment (feedmypixel) [reply src/apps/interactions/router.js]: > Users don't care about url's though
@ztolley agreed, but that is a different issue to the one we are discussing here.

File: src/apps/interactions/repos.js [hunk 17]
```diff
@@ -1,7 +1,7 @@
 const config = require('../../../config')
 const authorisedRequest = require('../../lib/authorised-request')
 
-function fetchInteraction (token, interactionId) {
+function getInteraction (token, interactionId) {
```
Comment (teneightfive) [inline src/apps/interactions/repos.js]: Maybe revert this change for now until we come up with a pattern that we replicate across the codebase. Feels like its adding noise to this change.
